File: bindings/julia/Treeweave/examples/zeta_bench.jl
# zeta_bench.jl — treeweave vs a fair brute-force Riemann-zeta eval.
# See examples/c++/zeta_bench.cpp for the rationale. ζ(s) = Σ_k k^-s summed until
# the tail is negligible (rel 1e-10, ≤160 terms) yet smooth on [2,10]: fit once.
# Times single/multi/sorted; the native rate is sampled over n_native and reused.
# TREEWEAVE_BENCH_YAML=path emits YAML.
using Treeweave
using Printf

# Fair baseline: sum k^-s until a term is below ZETA_EPS relative to the running
# total, capped at ZETA_MAX_TERMS — a competent zeta stops early.
const ZETA_EPS = 1e-10
const ZETA_MAX_TERMS = 160

# ζ(s) ≈ Σ_k k^-s (early stop). Both the fit callback and the native baseline.
function zeta_partial(s)
    acc = 0.0
    for k in 1:ZETA_MAX_TERMS
        term = Float64(k)^(-s)
        acc += term
        term < ZETA_EPS * acc && break
    end
    return acc
end

# Scalar accumulation in a function (not global scope) so boxing doesn't dominate.
function scalar_sum(f, xs)
    s = 0.0
    for x in xs
        s += f(x)
    end
    return s
end

# One eval-mode YAML block. "%.17e" carries a dot, so YAML 1.1 reads a float.
function emit_block(io, name, tw, nat)
    @printf(io, "%s:\n", name)
    @printf(io, "  treeweave_mevals_s: %.17e\n", tw)
    @printf(io, "  native_mevals_s: %.17e\n", nat)
    @printf(io, "  speedup: %.17e\n", tw / nat)
end

const a, b = 2.0, 10.0
approx = fit(zeta_partial, a, b, 1e-10)  # default tol_kind is RelativeMax
println(approx)

n        = 1_000_000   # batch / sorted points
n_scalar = 100_000     # scalar-API points
n_native = 256         # brute-force sample (<=160 powers each)
xs        = a .+ (b - a) .* rand(n)
xs_sorted = sort(xs)

# --- accuracy vs the brute-force sum, on the n_native sample -----------------
xs_native = xs[1:n_native]
yhat    = approx(xs_native)
yref    = zeta_partial.(xs_native)
max_rel = maximum(abs.(yhat .- yref) ./ abs.(yref))

mevals(count, seconds) = count / (seconds * 1e6)

# --- native rate: brute-force sum over the small sample (mode-independent) ---
scalar_sum(zeta_partial, xs_native)                                # warm-up (compile)
t0 = time_ns(); s_nat = scalar_sum(zeta_partial, xs_native); t1 = time_ns()
nat_s    = (t1 - t0) / 1e9
nat_rate = mevals(n_native, nat_s)                                 # Mevals/s, all modes
@assert isfinite(s_nat)

# --- single-eval: the scalar API, one point at a time ------------------------
xs_scalar = xs[1:n_scalar]
scalar_sum(approx, xs_scalar)                                      # warm-up (compile)
t0 = time_ns(); s_tw = scalar_sum(approx, xs_scalar); t1 = time_ns()
tw_single_s = (t1 - t0) / 1e9
@assert isfinite(s_tw)

# --- multi-eval: the unsorted batch (in place, allocation-free) --------------
tw_buf = similar(xs)
approx(xs; out=tw_buf)                                             # warm-up
t0 = time_ns(); approx(xs; out=tw_buf); t1 = time_ns()
tw_multi_s = (t1 - t0) / 1e9
@assert isfinite(sum(tw_buf))

# --- sorted-eval: the 1-D ascending fast path --------------------------------
approx(xs_sorted; sorted=true, out=tw_buf)                        # warm-up
t0 = time_ns(); approx(xs_sorted; sorted=true, out=tw_buf); t1 = time_ns()
tw_sorted_s = (t1 - t0) / 1e9
@assert isfinite(sum(tw_buf))

# --- throughput (Mevals/s) and speedup per mode ------------------------------
tw_single = mevals(n_scalar, tw_single_s)
tw_multi  = mevals(n, tw_multi_s)
tw_sorted = mevals(n, tw_sorted_s)

@printf("zeta(s) = sum_k k^-s (<=%d terms, stop at %.0e rel), fit on [%.1f, %.1f], relative tol %.0e\n",
        ZETA_MAX_TERMS, ZETA_EPS, a, b, 1e-10)
@printf("  max rel err: %.3e\n", max_rel)
@printf("  single-eval  treeweave %.1f  native %.4f Mevals/s  speedup %.1fx\n", tw_single, nat_rate, tw_single / nat_rate)
@printf("  multi-eval   treeweave %.1f  native %.4f Mevals/s  speedup %.1fx\n", tw_multi, nat_rate, tw_multi / nat_rate)
@printf("  sorted-eval  treeweave %.1f  native %.4f Mevals/s  speedup %.1fx\n", tw_sorted, nat_rate, tw_sorted / nat_rate)

# --- machine-readable YAML (optional) ----------------------------------------
yaml_path = get(ENV, "TREEWEAVE_BENCH_YAML", "")
if !isempty(yaml_path)
    open(yaml_path, "w") do io
        @printf(io, "language: \"julia\"\n")
        @printf(io, "domain: [%.17e, %.17e]\n", a, b)
        @printf(io, "tol: %.17e\n", 1e-10)
        @printf(io, "n_pts: %d\n", n)
        @printf(io, "max_rel_err: %.17e\n", max_rel)
        emit_block(io, "single_eval", tw_single, nat_rate)
        emit_block(io, "multi_eval",  tw_multi,  nat_rate)
        emit_block(io, "sorted_eval", tw_sorted, nat_rate)
    end
end

File: bindings/fortran/benchmarks/zeta_bench.f90
! zeta_bench.f90: treeweave vs a fair brute-force Riemann-zeta eval, via the C ABI.
! See benchmarks/zeta_bench.cpp for the rationale. ζ(s) = Σ_k k^-s summed until
! the tail is negligible (rel 1e-10, ≤160 terms) yet smooth on [2,10]: fit once.
! Times single/multi/sorted; the native rate is sampled over n_native and reused.
! TREEWEAVE_BENCH_YAML=path emits YAML.

module zeta_bench_kernels
    use, intrinsic :: iso_c_binding
    implicit none
    ! Fair baseline: sum k^-s until a term is below zeta_eps relative to the
    ! running total, capped at zeta_max_terms, a competent zeta stops early.
    real(c_double), parameter :: zeta_eps = 1.0e-10_c_double
    integer,        parameter :: zeta_max_terms = 160
contains
    ! ζ(s) ≈ Σ_k k^-s (early stop). Used both as the treeweave fit callback and
    ! as the native baseline (apples-to-apples).
    function zeta_partial(s) result(acc)
        real(c_double), intent(in) :: s
        real(c_double)             :: acc, term
        integer                    :: k
        acc = 0.0_c_double
        do k = 1, zeta_max_terms
            term = real(k, c_double)**(-s)
            acc = acc + term
            if (term < zeta_eps * acc) exit
        end do
    end function zeta_partial

    subroutine kernel_zeta(x, y, context) bind(C)
        real(c_double), intent(in)  :: x(*)
        real(c_double), intent(out) :: y(*)
        type(c_ptr),    value       :: context
        y(1) = zeta_partial(x(1))
    end subroutine kernel_zeta

    ! In-place ascending quicksort (Hoare partition), builds the sorted input
    ! for the sorted-eval mode. The sample is random, so recursion depth stays
    ! O(log n); the sort itself is untimed.
    recursive subroutine quicksort(arr, lo, hi)
        real(c_double), intent(inout) :: arr(:)
        integer,        intent(in)    :: lo, hi
        integer        :: i, j
        real(c_double) :: pivot, tmp
        if (lo >= hi) return
        pivot = arr((lo + hi) / 2)
        i = lo
        j = hi
        do
            do while (arr(i) < pivot)
                i = i + 1
            end do
            do while (arr(j) > pivot)
                j = j - 1
            end do
            if (i <= j) then
                tmp = arr(i); arr(i) = arr(j); arr(j) = tmp
                i = i + 1
                j = j - 1
            end if
            if (i > j) exit
        end do
        if (lo < j) call quicksort(arr, lo, j)
        if (i < hi) call quicksort(arr, i, hi)
    end subroutine quicksort

    ! Write one eval-mode block of the YAML document to unit u. The ES edit
    ! descriptor always emits "d.dddE±dd" (leading dot + exponent), so a YAML 1.1
    ! parser reads each value as a float rather than a string.
    subroutine emit_block(u, name, tw, nat)
        integer,          intent(in) :: u
        character(len=*), intent(in) :: name
        real(c_double),   intent(in) :: tw, nat
        write (u, '(A,A)')       name, ':'
        write (u, '(A,ES24.16)') '  treeweave_mevals_s: ', tw
        write (u, '(A,ES24.16)') '  native_mevals_s: ', nat
        write (u, '(A,ES24.16)') '  speedup: ', tw / nat
    end subroutine emit_block
end module zeta_bench_kernels


program zeta_bench
    use, intrinsic :: iso_c_binding
    use treeweave
    use zeta_bench_kernels
    implicit none

    integer, parameter          :: n = 1000000        ! batch / sorted points
    integer, parameter          :: n_scalar = 100000  ! scalar-API points
    integer, parameter          :: n_native = 256     ! brute-force sample (<=160 powers each)
    type(c_ptr)                 :: fn
    real(c_double)              :: a(1), b(1), tol
    real(c_double), allocatable :: xs(:), res(:), xs_sorted(:)
    real(c_double)              :: approx, exact, rel, max_rel
    real(c_double)              :: nat_s, nat_rate
    real(c_double)              :: tw_single_s, tw_multi_s, tw_sorted_s
    real(c_double)              :: tw_single, tw_multi, tw_sorted
    real(c_double), volatile    :: sink          ! defeats dead-code elimination
    integer(c_int64_t)          :: c0, c1, rate
    integer                     :: i, u, ios, yaml_len
    character(len=4096)         :: yaml_path

    a(1) = 2.0_c_double
    b(1) = 10.0_c_double
    tol  = 1.0e-10_c_double

    ! c_null_ptr opts => defaults, whose tol_kind is RELATIVE_MAX (the right
    ! measure for this zero-free, monotone function).
    fn = treeweave_fit(c_funloc(kernel_zeta), 1_c_int, 1_c_int, a, b, tol, &
                      c_null_ptr, c_null_ptr)
    if (.not. c_associated(fn)) then
        write (*, '(2A)') "treeweave_fit failed: ", treeweave_error_message()
        error stop 1
    end if

    allocate (xs(n), res(n), xs_sorted(n))
    call random_seed()
    call random_number(xs)
    xs = a(1) + (b(1) - a(1)) * xs

    call system_clock(count_rate=rate)

    max_rel = 0.0_c_double
    do i = 1, n_native
        exact  = zeta_partial(xs(i))
        approx = treeweave_eval_1d(fn, xs(i))
        rel    = abs(approx - exact) / abs(exact)
        if (rel > max_rel) max_rel = rel
    end do

    sink = 0.0_c_double

    ! --- native rate: brute-force sum over the small sample (mode-independent).
    ! Measured once and reused as the baseline in all three eval modes.
    do i = 1, n_native
        sink = sink + zeta_partial(xs(i))            ! warm-up (untimed)
    end do
    call system_clock(c0)
    do i = 1, n_native
        sink = sink + zeta_partial(xs(i))
    end do
    call system_clock(c1)
    nat_s    = real(c1 - c0, c_double) / real(rate, c_double)
    nat_rate = real(n_native, c_double) / (nat_s * 1.0e6_c_double)   ! Mevals/s, all modes

    do i = 1, n_scalar
        sink = sink + treeweave_eval_1d(fn, xs(i))    ! warm-up (untimed)
    end do
    call system_clock(c0)
    do i = 1, n_scalar
        sink = sink + treeweave_eval_1d(fn, xs(i))
    end do
    call system_clock(c1)
    tw_single_s = real(c1 - c0, c_double) / real(rate, c_double)

    call treeweave_batch(fn, xs, res, int(n, c_size_t))   ! warm-up (untimed)
    sink = res(1)
    call system_clock(c0)
    call treeweave_batch(fn, xs, res, int(n, c_size_t))
    call system_clock(c1)
    tw_multi_s = real(c1 - c0, c_double) / real(rate, c_double)
    sink = res(1)

    ! --- sorted-eval: the 1-D ascending fast path --------------------------
    ! Sort once, untimed.
    xs_sorted = xs
    call quicksort(xs_sorted, 1, n)
    call treeweave_sorted(fn, xs_sorted, res, int(n, c_size_t))   ! warm-up (untimed)
    sink = res(1)
    call system_clock(c0)
    call treeweave_sorted(fn, xs_sorted, res, int(n, c_size_t))
    call system_clock(c1)
    tw_sorted_s = real(c1 - c0, c_double) / real(rate, c_double)
    sink = res(1)

    tw_single = real(n_scalar, c_double) / (tw_single_s * 1.0e6_c_double)
    tw_multi  = real(n, c_double) / (tw_multi_s * 1.0e6_c_double)
    tw_sorted = real(n, c_double) / (tw_sorted_s * 1.0e6_c_double)

    write (*, '(A,I0,A,F0.1,A,F0.1,A,ES8.1)') "zeta(s) = sum k^-s (<=", zeta_max_terms, &
        " terms, stop at 1e-10), fit on [", a(1), ", ", b(1), "], relative tol ", tol
    write (*, '(A,ES10.3)') "  max rel err: ", max_rel
    write (*, '(A,F0.1,A,F0.4,A,F0.1,A)') "  single-eval  treeweave ", tw_single, "  native ", nat_rate, &
        " Mevals/s  speedup ", tw_single / nat_rate, "x"
    write (*, '(A,F0.1,A,F0.4,A,F0.1,A)') "  multi-eval   treeweave ", tw_multi, "  native ", nat_rate, &
        " Mevals/s  speedup ", tw_multi / nat_rate, "x"
    write (*, '(A,F0.1,A,F0.4,A,F0.1,A)') "  sorted-eval  treeweave ", tw_sorted, "  native ", nat_rate, &
        " Mevals/s  speedup ", tw_sorted / nat_rate, "x"

    call get_environment_variable("TREEWEAVE_BENCH_YAML", yaml_path, yaml_len)
    if (yaml_len > 0) then
        open (newunit=u, file=trim(yaml_path), status="replace", action="write", iostat=ios)
        if (ios == 0) then
            write (u, '(A)')                     'language: "fortran"'
            write (u, '(A,ES24.16,A,ES24.16,A)') 'domain: [', a(1), ', ', b(1), ']'
            write (u, '(A,ES24.16)')             'tol: ', tol
            write (u, '(A,I0)')                  'n_pts: ', n
            write (u, '(A,ES24.16)')             'max_rel_err: ', max_rel
            call emit_block(u, 'single_eval', tw_single, nat_rate)
            call emit_block(u, 'multi_eval', tw_multi, nat_rate)
            call emit_block(u, 'sorted_eval', tw_sorted, nat_rate)
            close (u)
        end if
    end if

    deallocate (xs, res, xs_sorted)
    fn = treeweave_free(fn)

    if (max_rel >= 1.0e-7_c_double) error stop 2
end program zeta_bench
